myokit/formats/diffsl/_exporter.py

```diff
@@ -0,0 +1,523 @@
+#
+# Export as a DiffSL model.
+#
+# This file is part of Myokit.
+# See http://myokit.org for copyright, sharing, and licensing details.
+#
+import collections
+import os
+import warnings
+
+import myokit
+import myokit.formats
+import myokit.formats.diffsl as diffsl
+import myokit.lib.guess as guess
+import myokit.lib.hh as hh
+import myokit.lib.markov as markov
+
+
+class DiffSLExporter(myokit.formats.Exporter):
+    """
+    This :class:`Exporter <myokit.formats.Exporter>` generates a DiffSL
+    implementation of a Myokit model.
+
+    Only the model definition is exported. All state variables and currents
+    are output, but no inputs are provided.
+
+    For details of the language, see https://martinjrobins.github.io/diffsl/
+    """
+
+    def model(self, path, model, protocol=None, convert_units=True):
+        """
+        Exports a :class:`myokit.Model` in DiffSL format, writing the result to
+        the file indicated by ``path``.
+
+        A :class:`myokit.ExportError` will be raised if any errors occur.
+
+        Arguments:
+
+        ``path``
+            The path to write the generated model to.
+        ``model``
+            The :class:`myokit.Model` to export.
+        ``protocol``
+            Not implemented!
+        ``convert_units``
+            If set to ``True``, the method will attempt to convert to
+            preferred units for voltage (mV), current (A/F), and time (ms).
+        """
+        # Raise exception if path is unwritable
+        self._test_writable_dir(os.path.dirname(path))
```

Comment:
I see your point. I moved it here because I thought it would not be worth going through the trouble to convert the model if you can't write it in the end. I'll modify it to check this at write time instead.

Reply:
Resolved by 1a21249